src/include/enclave/ticketlock.h

```diff
@@ -1,55 +1,111 @@
 #ifndef TICKETLOCK_H
 #define TICKETLOCK_H
 
-#include "atomic.h"
+#include <stdatomic.h>
 
 #if DEBUG
 #include "lthread.h"
 #endif /* DEBUG */
 
+/**
+ * A ticket lock.  This provides a simple spinlock that guarantees that the
+ * lock will be acquired in the order that callers attempt to acquire it.
+ *
+ * This is similar to the mechanism used in some shops, where customers take a
+ * numbered ticket and wait until that number is called.  Threads that attempt
+ * to acquire the lock 'take a ticket' by atomically incrementing one counter
+ * and then wait until that ticket is the current one.  Threads releasing the
+ * lock increment a second counter representing the current ticket.
+ */
 struct ticketlock
 {
+    /**
+     * Anonymous union of the two words used to represent the ticket lock.
+     * It's not clear whether doing different-sized atomic accesses to the same
+     * address is defined behaviour in C.  It is definitely broken on some
+     * PowerPC implementations but is fine on x86.
+     */
     union {
-        uint64_t u;
+        /**
+         * A single view of the ticket and users fields.  This is needed for
+         * `ticket_trylock`, which attempts to atomically test `ticket` and set
+         * `users`.
+         */
+        _Atomic(uint64_t) u;
         struct
         {
-            uint32_t ticket;
-            uint32_t users;
+            /**
+             * The current ticket value.  Each thread that attempts to acquire
+             * a ticket lock waits until this field matches its ticket.
+             */
+            _Atomic(uint32_t) ticket;
+            /**
+             * The value that the next thread attempting to acquire the lock
+             * should take.  A thread 'takes a ticket' by atomically
+             * incrementing this field.  The value before the increment is the
+             * value of that thread's ticket.
+             */
+            _Atomic(uint32_t) users;
         } s;
     };
 #if DEBUG
     struct lthread* lt; // Thread that is holding the lock.
 #endif                  /* DEBUG */
 };
 
+/**
+ * Acquire a ticket lock.  This spins until the lock is acquired, it does *not*
+ * yield to other cooperatively scheduled threads.
+ */
 static inline void ticket_lock(struct ticketlock* t)
 {
-    uint32_t me = a_fetch_add_uint(&t->s.users, 1);
+    // Acquire a ticket
+    uint32_t me = atomic_fetch_add(&t->s.users, 1);
+    // Wait until our ticket is called.
     while (t->s.ticket != me)
-        a_spin();
+    {
+#ifdef __x86__
+        __builtin_ia32_pause()
+#endif
+            ;
+    }
+
 #if DEBUG
     t->lt = NULL;
     // t->lt = lthread_self();
 #endif /* DEBUG */
 }
 
+/**
+ * Release the ticket lock.
+ */
 static inline void ticket_unlock(struct ticketlock* t)
 {
-    a_barrier();
```

Comment:
Not sure if this is a valid concern, but theres nothing guaranteeing if the thread attempting the unlock actually holds the lock.

Reply:
This is because the lock owner is sometimes not an lthread. Some of the invocations are from the lthread scheduler directly, with no running lthread.